src/lib.rs

```diff
@@ -99,6 +100,112 @@ where
+const EBDA_END: usize = 0x9ffff;
+/// The start of the main bios area below 1mb in which to search for the RSDP
+/// (Root System Description Pointer)
+const RSDP_BIOS_AREA_START: usize = 0xe0000;
+/// The end of the main bios area below 1mb in which to search for the RSDP
+/// (Root System Description Pointer)
+const RSDP_BIOS_AREA_END: usize = 0xfffff;
+/// The RSDP (Root System Description Pointer)'s signature, "RSD PTR " (note trailing space)
+const RSDP_SIGNATURE: &'static [u8; 8] = b"RSD PTR ";
+
+/// Find the begining of the EBDA (Extended Bios Data Area) and return `None` if the ptr at
+/// `0x40e` is invalid.
+fn find_ebda_start<H>(handler: &mut H) -> Option<usize>
+where
+    H: AcpiHandler,
+{
+    // Read base segment from BIOS area. This is not always given by the bios, so it needs to be
+    // checked. We left shift 4 because it is a segment ptr.
+    let base_mapping = handler.map_physical_region::<u16>(
+        EBDA_START_SEGMENT_PTR, mem::size_of::<u16>()
+    );
+    let base = (*base_mapping as usize) << 4;
+    handler.unmap_physical_region(base_mapping);
+
+    // Check if base segment ptr is in valid range valid
+    if (EBDA_EARLIEST_START..EBDA_END).contains(&base) {
+        debug!("EBDA address is {:#x}", base);
+        Some(base)
+    } else {
+        warn!(
+            "EBDA address at {:#x} out of range ({:#x}), falling back to {:#x}",
+            EBDA_START_SEGMENT_PTR,
+            base,
+            EBDA_EARLIEST_START
+        );
+
+        None
+    }
+}
+
+/// This is the entry point of `acpi` if you have no information except that the machine is running
+/// BIOS and not UEFI. It maps the RSDP, works out what version of ACPI the hardware supports, and
+/// passes the physical address of the RSDT/XSDT to `parse_rsdt`.
+///
+/// # Unsafety
+///
+/// This function is unsafe because it may read from protected memory if the computer is using UEFI.
+/// Only use this function if you are sure the computer is using BIOS.
+pub unsafe fn search_for_rsdp_bios<H>(handler: &mut H) -> Result<(), AcpiError>
+where
+    H: AcpiHandler,
+{
+    let ebda_start = find_ebda_start(handler);
+
+    // The areas that will be searched for the RSDP
+    let areas = [
+        // Main bios area below 1 mb
+        // In practice (from my [Restioson's] testing, at least), the RSDP is more often here than
+        // the in EBDA. Also, if we cannot find the EBDA, then we don't want to search the largest
+        // possible EBDA first.
+        RSDP_BIOS_AREA_START..=RSDP_BIOS_AREA_END,
+
+        if let Some(ebda_start) = ebda_start {
+            // First kb of EBDA
+            ebda_start..=ebda_start + 1024
+        } else {
+            // We don't know where the EBDA starts, so just search the largest possible EBDA
+            EBDA_EARLIEST_START..=EBDA_END
+        }
+    ];
+
+    // Signature is always on a 16 byte boundary so only search there
+    for address in areas.into_iter().flat_map(|i| i.clone()).step_by(16) {
```

Comment:
Why does `i` need to be cloned? Should `i` not be owned at this stage?

Reply:
I think I did try this. You can try removing it though.